swim/heal_partition_test.go

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/stretchr/testify/assert"
+	"github.com/uber/ringpop-go/discovery/statichosts"
+)
+
+func TestPartitionHealWithFaulties(t *testing.T) {
+	A := newPartition(t, 5)
+	B := newPartition(t, 5)
+	defer destroyNodes(A...)
+	defer destroyNodes(B...)
+
+	A.AddPartitionWithStatus(B, Faulty)
+	B.AddPartitionWithStatus(A, Faulty)
+
+	A.ProgressTime(time.Millisecond * 3)
+	B.ProgressTime(time.Millisecond * 5)
+
+	A[0].node.discoverProvider = statichosts.New(append(A.Hosts(), B.Hosts()...)...)
+
+	targets := A[0].node.healer.Heal()
+	assert.Len(t, targets, 1, "expected correct amount of targets")
+	assert.True(t, B.Contains(targets[0]), "expected target to be a node from the right partition")
+
+	waitForConvergence(t, 3*time.Second, A...)
+	waitForConvergence(t, 3*time.Second, B...)
+
+	A.HasPartitionAs(t, A, 3, "alive")
+	A.HasPartitionAs(t, B, 0, "faulty")
+	B.HasPartitionAs(t, B, 5, "alive")
+	B.HasPartitionAs(t, A, 0, "faulty")
+
+	targets = A[0].node.healer.Heal()
+	assert.Len(t, targets, 1, "expected correct amount of targets")
+	assert.True(t, B.Contains(targets[0]), "expected target to be a node from the right partition")
+
+	waitForPartitionHeal(t, time.Second*3, A, B)
+
+	A.HasPartitionAs(t, A, 3, "alive")
+	A.HasPartitionAs(t, B, 5, "alive")
+	B.HasPartitionAs(t, B, 5, "alive")
+	B.HasPartitionAs(t, A, 3, "alive")
+}
+
+func TestPartitionHealWithMissing(t *testing.T) {
```

Comment:
Also, a high-level description of the test would be nice (in the similar spirit like I wrote above).